### PR TITLE
[Example] add generation inference script for deepspeed

### PR DESCRIPTION
**Pull request by dc3671**
## Type of Change

add launch script for generation with deepspeed.

## Description

example for run inference:
```
mpirun -np 2 --prepend-rank python -u run_generation_with_deepspeed.py -m EleutherAI/gpt-j-6b --benchmark --ipex --input-tokens=1024 --max-new-tokens=128
```

example for run accuracy test:
```
LLM_ACC_TEST=1 mpirun -np 2 --prepend-rank python -u run_generation_with_deepspeed.py -m EleutherAI/gpt-j-6b --accuracy-only --ipex
```

@changwangss  @hshen14 Could you please take a look at this PR? We may need this script in public access. Thanks~ 

cc @rogerxfeng8

### CODE REVIEW

**Comment by hshen14** [comment]
CC @ftian1 @lvliang-intel for review

File: examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+            return val
+    return default
+
+local_rank = get_int_from_env(["LOCAL_RANK","MPI_LOCALRANKID","PALS_LOCAL_RANKID"], "0")
+world_size = get_int_from_env(["WORLD_SIZE","PMI_SIZE","PALS_LOCAL_SIZE"], "1")
+port = get_int_from_env(["MASTER_PORT"], 29500)
+print(f"*** local_rank={local_rank} world_size={world_size}")
+
+deepspeed.init_distributed(get_accelerator().communication_backend_name(), distributed_port=port)
+x = torch.ones(
```
Comment (ftian1) [inline examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py:125]: may I know what's it for? (line 125 to line 128)

File: examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+    model,
+    mp_size=world_size,
+    base_dir=repo_root,
+    dtype=infer_dtype,
+    checkpoint=checkpoints_json if is_meta_support else None,
+    **kwargs,
+)
+
+if args.benchmark:
+    get_accelerator().empty_cache()
```
Comment (ftian1) [inline examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py:275]: these 3 lines are frequently used, suggest to be a util func for caller use.

**Comment by ftian1** [comment]
@dc3671  besides above comments I made, will you add a README to tell user how to run it on multiple nodes?

and can you change below cmd to use deepspeed launcher which is the superset of mpi.

```
mpirun -np 2 --prepend-rank python -u run_generation_with_deepspeed.py -m EleutherAI/gpt-j-6b --benchmark --ipex --input-tokens=1024 --max-new-tokens=128
```

File: examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    if model_class[0] == AutoModelForCausalLM and is_meta_support:
+        model = model_class[0].from_config(config, torch_dtype=load_dtype, trust_remote_code=True)
+    else:
+        model = model_class[0].from_pretrained(model_name, config=config, low_cpu_mem_usage=True, torch_dtype=load_dtype, trust_remote_code=True)
+
+if args.benchmark:
+    deepspeed.runtime.utils.see_memory_usage("post-from-pretrained", force=True)
+
+model = model.eval()
+model = model.to(memory_format=torch.channels_last)
```
Comment (ftian1) [inline examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py:227]: I remember channel last is specially optimized by IPEX but not cuda. I would suggest to do this line only if ipex is found or enabled.

File: examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py (hunk 1 above)
Comment (daisyden) [reply examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py:125]: Is it a debug code?

**Comment by daisyden** [comment]
deepspeed is the recommend tool to launch the program.
For single node multiple ranks you can refer to this command, it use the built-in launcher of deepspeed.
unset KMP_AFFINITY
deepspeed  --num_gpus 2 --master_addr `hostname -I | sed -e 's/\s.*$//'` --bind_cores_to_rank --bind_core_list 0-$((OMP_NUM_THREADS - 1)),56-$((56 + OMP_NUM_THREADS - 1)) distributed/run_generation_with_deepspeed.py  .....

For multinode it support different mpi launchers and can help you hide the compatibilities of arguments.
deepspeed --no_ssh_check --hostfile=nodefile  --bind_cores_to_rank --bind_core_list 0-$((OMP_NUM_THREADS - 1)),56-$((56 + OMP_NUM_THREADS - 1)) --launcher impi --launcher_args "--genv LD_LIBRARY_PATH=${LD_LIBRARY_PATH} $margs -l " distributed/run_generation_with_deepspeed.py  .....

File: examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py (hunk 1 above)
Comment (dc3671) [reply examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py:123]: Hi, I added a comment for this. It's kept for testing ccl allreduce ahead of time.

File: examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py (hunk 2 above)
Comment (dc3671) [reply examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py]: done

**Comment by dc3671** [comment]
> @dc3671 besides above comments I made, will you add a README to tell user how to run it on multiple nodes?
> 
> and can you change below cmd to use deepspeed launcher which is the superset of mpi.
> 
> ```
> mpirun -np 2 --prepend-rank python -u run_generation_with_deepspeed.py -m EleutherAI/gpt-j-6b --benchmark --ipex --input-tokens=1024 --max-new-tokens=128
> ```

Hi @ftian1 , I talked with Daisy and I think in XPU inference scenario, we need to use mpi launcher for better performance. I tried to use deepspeed launcher, but it requires `CCL_ZE_IPC_EXCHANGE=sockets` which we don't want.

File: examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py (hunk 3 above)
Comment (dc3671) [reply examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py]: I removed this.

**Comment by ftian1** [comment]
> > @dc3671 besides above comments I made, will you add a README to tell user how to run it on multiple nodes?
> > and can you change below cmd to use deepspeed launcher which is the superset of mpi.
> > ```
> > mpirun -np 2 --prepend-rank python -u run_generation_with_deepspeed.py -m EleutherAI/gpt-j-6b --benchmark --ipex --input-tokens=1024 --max-new-tokens=128
> > ```
> 
> Hi @ftian1 , I talked with Daisy and I think in XPU inference scenario, we need to use mpi launcher for better performance. I tried to use deepspeed launcher, but it requires `CCL_ZE_IPC_EXCHANGE=sockets` which we don't want.

I would suggest to add both ways in README/code for user to choose.

**Review by ftian1 — APPROVED** [review]
The PR looks good to me. but I suggest to keep mpi and add 'deepspeed' launcher back to be compatible with office doc.

**Review by daisyden — APPROVED** [review]

**Review by ftian1 — APPROVED** [review]
